### DISCUSSION TITLE
mark type "shape"

### DISCUSSION THREAD

**Discussion by flekschas** [gosling-lang/gosling.js#241] [Ideas]
I saw in the documentation the mark type "triangle" and that made me wonder if there could be an option for a "generic" mark type called maybe `shape`, which allows the user to specify a shape instance that they would like to place at specific locations in the genome.

For instance, _iobio_ (https://bivi.co/visualisation/iobio) has this track where different point events are represented by different shapes. A triangle could be one of this shape but there could be others.

![gene iobio_](https://user-images.githubusercontent.com/932103/108549677-7ad28580-72bb-11eb-9e3c-85642854dedb.png)

Specification-wise I could think of an approach similar to how SVG handles `symbol`: https://developer.mozilla.org/en-US/docs/Web/SVG/Element/symbol. The user could define an shape in the "header" of the specification and then later on use it as a mark.

Brainstorming this further we could even think of this as a `markDefinition` that can later on be used via `"mark": "markDefinitionName"`.

The biggest question of course is how to optimize this for WebGL rendering. One approach could be similar to how Piling.js first rasterizes arbitrary SVG into an image and then uses that image as a texture object in WebGL.

**Comment by sehilyi** [comment]
Interesting to see so diverse shapes being used in the genomic-related tool!

I agree that generic mark types (`shape`) and custom marks can be useful for some cases, especially when people want to represent many categories and even use semantic shapes (e.g., first-aid kit for pathogenicity) for better interpretation.

Perhaps, when we start implementing this, we can refer to the list of generic marks that are supported in the `vega-lite`'s `shape` mark.

Btw, can you share me the link to your code for the texture reusing part?

**Comment by sehilyi** [comment]
Maybe getting too far from your original idea, but perhaps it would be useful to even allow defining `glyph` where its visual properties can be bound to data (e.g., size and color).

(I once tried to develop this idea to allow creating custom glyphs, something like `mark: { type: 'glyph', name: 'higlass-style-gene', elements: [...] }}` where each `element` specifies what generic mark type to use and how its visual properties should be bound to data. I imagined of using the similar approach taken in the "Glyph" panel on the left-top corner of Charticulator (https://charticulator.com/). But, I ended up simply allowing people to use the current `"overlay: []"` prop though.)
